bpo-40636: Documentation for zip-strict (#20961)
python/cpython: Doc/library/functions.rst

   dictionary are ignored.

.. function:: zip(*iterables, strict=False)

>    Invert a series of iterables, turning rows into columns, and columns into rows.

Reviewer: I don't think of zip() as "inverting a matrix". That sounds like a very data science oriented way of thinking. I see it as iterating in parallel over several iterables and producing the results as a stream of tuples that contain the next item from each iterable. In particular, I never think of either the input or the output as "rows" and "columns".

Reviewer 2: Yeah, completely agree. I think the original summary was fine:

```suggestion
   Make an iterator that aggregates elements from each of the iterables.
```

Author: I'll put the rows-and-columns approach out of the summary, even though I like it. 

I don't like the original summary "Make an iterator that aggregates elements from each of the iterables." The word "aggregate" is so generic, that it doesn't add any information. To aggregate can mean to add as a sum, to take an average, to do a union, [and other actions](https://docs.djangoproject.com/en/3.0/topics/db/aggregation/#cheat-sheet).

Let me think about another summary.

Author: Okay, I used some of Guido's words for the summary. I do think it's better than the "aggregate" sentence. I included the rows-and-columns approach outside of the summary, because I think it'll make the penny drop for many people.

   Example::

>       >>> for item in zip(('a', 'b', 'c'), (1, 2, 3)):

Reviewer: I honestly would use lists instead of tuples for the input. Also, perhaps assign them to variables first:
```
first = ['one', 'two', 'three']
second = [1, 2, 3]
for item in zip(first, second):
    print(item)
```

Author: Compromise, lists but not variables? The example is bigger than I'd like as it is.

      ...     print(item)
      ...
      ('a', 1)
      ('b', 2)
      ('c', 3)

   More formally: :func:`zip` returns an iterator of tuples, where the *i*-th
   tuple contains the *i*-th element from each of the argument sequences or

>    iterables.

Reviewer: "argument sequences or iterables" -> "arguments" (or possibly "argument iterables").

Author: Done.

>    :func:`zip` returns a lazy iterable. The elements wouldn't be processed

Reviewer: If you want to use the word "lazy", I would say "`zip` is lazy" instead. The thing it returns is an iterator, and all iterators are lazy -- we don't need to say that.

Author: Will do.

   until the iterable is iterated on, e.g. by wrapping in a :class:`tuple`.

   One possible wrinkle is that the iterables passed to :func:`zip` could have
   different lengths; sometimes by design, and sometimes because of a bug in
   the code that prepared these iterables.  Python offers three different

>    approaches to dealing with this issue:

Reviewer: This paragraph makes it sound too much as if the case of mismatching lengths is a problem ("wrinkle", "issue"). It is not. There simply are different use cases.

Author: Fixed.

>    * By default, :func:`zip` has short-circuit behavior.  If any of the

Reviewer: That is not what "short-circuit" in programming means! Just say "By default, `zip` stops when the shortest iterable is exhausted."

Author: Done.

>      iterables are shorter than the rest, :func:`zip` will ignore the remainders

Reviewer 2: "remainders" sounds like math to me.

```suggestion
     iterables are shorter than the rest, :func:`zip` will ignore the remaining items
```

Author: Agreed, fixed.

     of the longer iterables, cutting off the result to the length of the shortest
     iterable::

        >>> tuple(zip(range(3), ['fee', 'fi', 'fo', 'fum']))

>         ((0, 'fee'), (1, 'fi'), (2, 'fo'))

Reviewer: I'd use `list(...)`. Also show the other case, `zip(range(10), ['fee', 'fi', 'fo', fum'])` (to clarify by example that it's not the length of the first iterable that determines the outcome).

Author: I've changed it to `list`, but I disagree about showing the other case. The entire topic of different lengths of iterables is by itself pretty esoteric, and it already gets a lot of words devoted to it in the `zip` documentation, so I want to avoid adding too much. If you feel strongly about this, I'm willing to let it go and include the other case.

Reviewer: Okay, that's fair. Just don't claim it's esoteric on python-ideas. :-)

>    * :func:`zip` is often used in cases where the iterables are assumed to be

Reviewer 2: ```suggestion
   * :func:`zip` is often used in cases where the lengths of the iterables are assumed to be
```

Author: Compromised.

     equal.  In such cases, it's recommended to use the ``strict=True`` option.
     Its output is the same as regular :func:`zip`::

        >>> tuple(zip(('a', 'b', 'c'), (1, 2, 3), strict=True))
        (('a', 1), ('b', 2), ('c', 3))

>      Unlike the default behavior, it asserts that the lengths of iterables is

Reviewer 2: ```suggestion
     Unlike the default behavior, it checks that the lengths of iterables is
```

Author: Fixed.

     identical, raising a :exc:`ValueError` if they aren't:

        >>> tuple(zip(range(3), ['fee', 'fi', 'fo', 'fum'], strict=True))
        ValueError: zip() argument 2 is longer than argument 1

     Without the ``strict=True`` argument, any bug that results in iterables of
     different lengths will be silenced, possibly mainfesting as a hard-to-find
     bug in another part of the program.

   * Shorter iterables can be padded with a constant value to make all the
     iterables equal. This is done by :func:`itertools.zip_longest`.

   Special cases: With a single iterable argument, :func:`zip` returns an

>    iterator of 1-tuples.  With no arguments, it returns an empty iterator.

Reviewer 2: These aren't special cases. I don't think this paragraph is needed.

Author: Funny, this paragraph was already there before I changed the documentation. I added the "Special cases:" heading because I do see these as edge cases. Would "Edge cases: " be more logical? Maybe "special" conveys that they have a behavior that's different than normal cases.

Reviewer 2: I personally wouldn't choose to include it, since it follows naturally from the described behavior. *Maybe* the zero-length case could be considered unspecified without this, but that's such an uncommon case with obvious behavior.

But if we keep it, I don't think we need to add "Edge cases:" or "Special cases:" or anything. It makes it seem more important than it really is.

Reviewer: Since this is user documentation, not a formal specification, I think it's nice to include what happens in edge cases even if it follows logically from the rest of the docs. And yes, "Edge cases" is the right heading.

   Tips and tricks:

   * The left-to-right evaluation order of the iterables is guaranteed. This
     makes possible an idiom for clustering a data series into n-length groups

>      using ``zip(*[iter(s)]*n)``.  This repeats the *same* iterator ``n`` times

Reviewer 2: ```suggestion
     using ``zip(*[iter(s)]*n, strict=True)``.  This repeats the *same* iterator ``n`` times
```

Author: Done.

     so that each output tuple has the result of ``n`` calls to the iterator.
     This has the effect of dividing the input into n-length chunks.

   * :func:`zip` in conjunction with the ``*`` operator can be used to unzip a
     list::

        >>> x = [1, 2, 3]
        >>> y = [4, 5, 6]
        >>> zipped = zip(x, y)

>         >>> list(zipped)

Reviewer: Please combine these two lines -- let's not mislead readers to think that `zipped` here is the zipped sequence (it's just the iterator).
```suggestion
        >>> list(zip(x, y))
```

Author: Done.

        [(1, 4), (2, 5), (3, 6)]
        >>> x2, y2 = zip(*zip(x, y))
        >>> x == list(x2) and y == list(y2)
        True

   .. versionchanged:: 3.10

>       The ``strict`` argument was added

Reviewer 2: ```suggestion
      Added the ``strict`` argument.
```

Author: Done.

   dictionary are ignored.

.. function:: zip(*iterables, strict=False)

   Iterate over several iterables in parallel, producing tuples with an item
   from each one.

   Example::

      >>> for item in zip(['a', 'b', 'c'], [1, 2, 3]):
      ...     print(item)
      ...
      ('a', 1)
      ('b', 2)
      ('c', 3)

>    More formally: :func:`zip` returns an iterator of tuples, where the *i*-th

Reviewer 2: You use <code>:func:\`zip\`</code> like 10 times throughout this section (basically every time you refer to `zip`).

And all it does is link back to this same section; my browser doesn't even move, so it feels like a broken link! I would consider just using <code>\`\`code()\`\`</code> formatting for all of them. For example:

```suggestion
   More formally: ``zip()`` returns an iterator of tuples, where the *i*-th
```

Author: I'm okay with making this change, but I worry that it wouldn't be consistent with the other sections of the documentation. @gvanrossum What do you think?

Reviewer: Leave it as is. This is a common phenomenon.

   Example::

>       >>> for item in zip(['a', 'b', 'c'], [1, 2, 3]):

Reviewer: I would still recommend using three small words rather than single-character strings.

Author: Done.

      ...     print(item)
      ...
      ('a', 1)
      ('b', 2)
      ('c', 3)

   More formally: :func:`zip` returns an iterator of tuples, where the *i*-th
   tuple contains the *i*-th element from each of the argument iterables.

>    :func:`zip` is lazy: The elements wouldn't be processed until the iterable

Reviewer: wouldn't -> won't (I think)

Author: Done.

>    is iterated on, e.g. by wrapping in a :class:`list`.

Reviewer: Maybe also mention a `for` loop as a way to iterate on the iterable? (Obviously we don't need to mention all ways, but using a `for` loop is more idiomatic than wrapping it in `list()`, except at the REPL when you just want to see how it works. :-)

Author: Done.

   One thing to consider is that the iterables passed to :func:`zip` could have
   different lengths; sometimes by design, and sometimes because of a bug in
   the code that prepared these iterables.  Python offers three different
   approaches to dealing with this issue:

>    * By default, :func:`zip` stops when the shortest iterable is exhausted.

Reviewer: Maybe also mention `strict=False`?

Author: Sounds esoteric to me, unless I'm misunderstanding what you mean. I guess there are cases where someone would pass a variable into `strict` so it'll be false on certain conditions, but that sounds really rare.

Reviewer: I just meant changing "By default," into "By default, or when `strict=False` is passed. (There was a user on the mailing list who had a decent reason to write this.) But on second thought I think we needn't change it here, it's implied by the description of `strict=True`.

>      :func:`zip` will ignore the remaining items in the longer iterables, cutting

Reviewer: Change "zip" to "it" here, to avoid repetition.

Author: Done.

   More formally: :func:`zip` returns an iterator of tuples, where the *i*-th
   tuple contains the *i*-th element from each of the argument iterables.

   Another way to think of :func:`zip` is that it turns rows into columns, and
   columns into rows.  This is similar to `transposing a matrix

>    <https://en.wikipedia.org/wiki/Transpose>`_.

Reviewer: I like it when you put it like this!

Author: :)
